spring-vault-core/src/test/java/org/springframework/vault/core/VaultTransitTemplateIntegrationTests.java

```diff
@@ -277,4 +282,126 @@ public void encryptAndRewrapShouldCreateCiphertextWithNonceAndContext() {
+	public void batchEncryptionAndDecryptionTestWithNonEqualContext() {
+
+		VaultTransitKeyCreationRequest request = VaultTransitKeyCreationRequest.builder() //
+				.derived(true) //
+				.build();
+
+		transitOperations.createKey("mykey", request);
+
+		List<String> plaintexts = new ArrayList<String>();
+		plaintexts.add("one");
+		plaintexts.add("two");
+
+		List<VaultTransitContext> encryptionContexts = new ArrayList<VaultTransitContext>();
+		encryptionContexts.add(VaultTransitContext.builder().context("oneContext".getBytes()).build());
+		encryptionContexts.add(VaultTransitContext.builder().context("twoContext".getBytes()).build());
+
+		List<VaultTransitContext> decryptionContext = new ArrayList<VaultTransitContext>();
+		decryptionContext.add(VaultTransitContext.builder().context("oneContext".getBytes()).build());
+
+		batchEncryptionAndDecryption(plaintexts, encryptionContexts, decryptionContext);
+	}
+
+	@Test(expected = VaultException.class)
+	public void batchEncryptionAndDecryptionTestWithNonMatchingContext() {
+
+		VaultTransitKeyCreationRequest request = VaultTransitKeyCreationRequest.builder() //
+				.derived(true) //
+				.build();
+
+		transitOperations.createKey("mykey", request);
+
+		List<String> plaintexts = new ArrayList<String>();
+		plaintexts.add("one");
+		plaintexts.add("two");
+
+		List<VaultTransitContext> encryptionContexts = new ArrayList<VaultTransitContext>();
+		encryptionContexts.add(VaultTransitContext.builder().context("oneContext".getBytes()).build());
+		encryptionContexts.add(VaultTransitContext.builder().context("twoContext".getBytes()).build());
+
+		List<VaultTransitContext> decryptionContext = new ArrayList<VaultTransitContext>();
+		decryptionContext.add(VaultTransitContext.builder().context("oneContext".getBytes()).build());
+		decryptionContext.add(VaultTransitContext.builder().context("wrongTwoContext".getBytes()).build());
+
+		batchEncryptionAndDecryption(plaintexts, encryptionContexts, decryptionContext);
+	}
+
+	private void batchEncryptionAndDecryption(List<String> plaintexts, List<VaultTransitContext> encryptionContexts,
+			List<VaultTransitContext> decryptionContext) {
+
+		List<Map<String, String>> cipherResponse = transitOperations.encrypt("mykey", plaintexts, encryptionContexts);
+
+		List<String> cipherList = new ArrayList<String>();
+
+		for (Map<String, String> entry : cipherResponse) {
+			cipherList.add(entry.get("ciphertext"));
+		}
+
+		List<Map<String, String>> plaintextResponse = transitOperations.decrypt("mykey", cipherList, decryptionContext);
+
+		/**
+		 * check to see if decryption ended up with errors. this is special
+		 * handling when context is supplied. For the other use cases, the
+		 * handling should be done by the applications.
+		 */
+		if (decryptionContext != null) {
+			for (Map<String, String> entry : plaintextResponse) {
+
+				if (!StringUtils.isEmpty(entry.get("error"))) {
+					throw new VaultException(entry.get("error"));
```

Comment:
That shouldn't happen here. It's the responsibility of the Template API implementation to handle errors and throw exceptions.

Reply:
originally I had this code inside the impl code. will put it back.

Reply:
In the light of a `EncryptionResult`/`DecryptionResult` we could align the design of the value objects to a `Future`. Result types can expose a `get()` method that either returns the result or throws an exception. We should make sure to provide introspection methods like `isSuccess()` and `getCause()` for a smooth usage.

Does this make sense?

Reply:
makes sense. very good suggestion.